make/compiler_flags

```diff
@@ -0,0 +1,191 @@
+# - CXX_MINOR: minor version of CXX
+ifneq (,$(findstring clang,$(CXX)))
+  CXX_TYPE ?= clang
+endif
+ifneq (,$(findstring mingw32-g,$(CXX)))
+  CXX_TYPE ?= mingw32-gcc
+endif
+ifneq (,$(findstring gcc,$(CXX)))
+  CXX_TYPE ?= gcc
+endif
+ifneq (,$(findstring g++,$(CXX)))
+  CXX_TYPE ?= gcc
+endif
+CXX_TYPE ?= other
+CXX_MAJOR := $(shell $(CXX) -dumpversion 2>&1 | cut -d'.' -f1)
+CXX_MINOR := $(shell $(CXX) -dumpversion 2>&1 | cut -d'.' -f2)
+
+################################################################################
+# Set default compiler flags
+#
+# The options that have been commented are things that the user can set.
+# They are commented because in make, undefined is different than empty;
+# we can test for that everywhere, but it's a lot easier using the set if
+# unassigned operator (`?=`) when it's not set.
+##
+O ?= 3
+
+## setup includes
+INC ?= $(INC_FIRST) -I $(if $(MATH),$(MATH),.) -I $(EIGEN) -I $(BOOST) $(INC_SUNDIALS)
+INC_SUNDIALS ?= -I $(SUNDIALS)/include
+INC_GTEST ?= -I $(GTEST)/include -I $(GTEST)
+
+## setup precompiler options
+CPPFLAGS_BOOST ?= -DBOOST_RESULT_OF_USE_TR1 -DBOOST_NO_DECLTYPE -DBOOST_DISABLE_ASSERTS -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION
+CPPFLAGS_SUNDIALS ?= -DNO_FPRINTF_OUTPUT
+CPPFLAGS_GTEST ?= -DGTEST_USE_OWN_TR1_TUPLE
+
+
+## setup compiler flags
+CXXFLAGS_LANG ?= -std=c++1y
+#CXXFLAGS_BOOST
+CXXFLAGS_SUNDIALS ?= -pipe
+#CXXFLAGS_GTEST
+
+
+## These are the other compiler flags that can be set.
+##
+## They are listed here for documentation and are intentionally
+## commented out.
+
+# CXXFLAGS_OS
+# CXXFLAGS_WARNINGS
+# CXXFLAGS_EIGEN
+
+## GPU specific flags
+# CXXFLAGS_OPENCL
+# LDFLAGS_OPENCL
+# LDLIBS_OPENCL
+
+## MPI specific flags
+# CXXFLAGS_MPI
+# LDFLAGS_MPI
+# LDLIBS_MPI
+
+################################################################################
+# Update compiler flags with operating system specific modifications
+##
+ifeq ($(OS),Windows_NT)
+  CXXFLAGS_WARNINGS ?= -Wall -Wno-unused-function -Wno-uninitialized -Wno-unused-but-set-variable -Wno-unused-variable -Wno-sign-compare -Wno-unused-local-typedefs
+  CPPFLAGS_GTEST ?= -DGTEST_HAS_PTHREAD=0
+  CPPFLAGS_OS ?= -D_USE_MATH_DEFINES
+  ## Detecting Process Bitness:
+  ## http://blogs.msdn.com/b/david.wang/archive/2006/03/26/howto-detect-process-bitness.aspx
+  ifeq ($(PROCESSOR_ARCHITECTURE)$(PROCESSOR_ARCHITEW6432),x86)
+    CXXFLAGS_OS ?= -m32
+  else
+    CXXFLAGS_OS ?= -m64
+  endif
+
+  ifneq (gcc,$(CXX_TYPE))
+    LDLIBS_OS ?= -static-libgcc
+  else
+    LDLIBS_OS ?= -static-libgcc -static-libstdc++
+  endif
+
+  EXE := .exe
+endif
+
+ifeq ($(OS),Darwin)
+  ifeq (clang,$(CXX_TYPE))
+    CXXFLAGS_OS ?= -Wno-unknown-warning-option -Wno-tautological-compare -Wno-sign-compare
+  endif
+endif
+
+ifeq ($(OS),Linux)
+  CPPFLAGS_GTEST ?= -DGTEST_HAS_PTHREAD=0
+  CXXFLAGS_WARNINGS ?= -Wno-sign-compare
+  ifeq (gcc,$(CXX_TYPE))
+    CXXFLAGS_OS ?= -pthread
+  endif
+  ifeq (mingw32-gcc,$(CXX_TYPE))
```

Comment:
Is this intentionally here? It looks odd to test under a detected Linux OS for mingw32-gcc stuff and switch on winpthreads. I am just flagging this at it looks odd.

Reply:
Yes, it is. This is something @bgoodri had put in and I'm just keeping the existing behavior. I believe it's for a cross-compiler.